pkg/virt-launcher/virtwrap/manager.go

```diff
@@ -580,6 +538,90 @@ func (l *LibvirtDomainManager) asyncMigrate(vmi *v1.VirtualMachineInstance, opti
 	}(l, vmi)
 }
 
+func setupMigration(l *LibvirtDomainManager, vmi *v1.VirtualMachineInstance, dom cli.VirDomain, options *cmdclient.MigrationOptions, loopbackAddress string) (*libvirt.DomainMigrateParameters, error) {
+	l.domainModifyLock.Lock()
+	defer l.domainModifyLock.Unlock()
+
+	if err := detachHostDevices(l.virConn, dom); err != nil {
+		log.Log.Object(vmi).Reason(err).Error(fmt.Sprintf("Live migration failed."))
+		return nil, err
+	}
+
+	params, err := prepareMigrationParams(options, vmi, loopbackAddress, dom)
+	if err != nil {
+		log.Log.Object(vmi).Reason(err).Error("Live migration failed.")
+		return nil, err
+	}
+	return params, nil
+}
+
+func prepareMigrationParams(options *cmdclient.MigrationOptions, vmi *v1.VirtualMachineInstance, loopbackAddress string, dom cli.VirDomain) (*libvirt.DomainMigrateParameters, error) {
+	bandwidth, err := converter.QuantityToMebiByte(options.Bandwidth)
+	if err != nil {
+		return nil, fmt.Errorf("invalid bandwidth supplied, err: %v", err)
+	}
+
+	migrURI := fmt.Sprintf("tcp://%s", ip.NormalizeIPAddress(loopbackAddress))
+
+	xmlstr, err := domXMLWithoutKubevirtMetadata(dom, vmi)
+	if err != nil {
+		return nil, fmt.Errorf("could not compute target XML, err: %v", err)
+	}
+
+	params := &libvirt.DomainMigrateParameters{
+		Bandwidth:  bandwidth, // MiB/s
+		URI:        migrURI,
+		URISet:     true,
+		DestXML:    xmlstr,
+		DestXMLSet: true,
+	}
+	copyDisks := getDiskTargetsForMigration(dom, vmi)
+	if len(copyDisks) != 0 {
+		params.MigrateDisks = copyDisks
+		params.MigrateDisksSet = true
+	}
+	return params, nil
+}
+
+func startMigrationProxyServer(virtShareDir string, vmi *v1.VirtualMachineInstance, loopbackAddress string) (stopServer func(), err error) {
+	var proxyStopListenPool []func()
+	stopServer = func() {
+		for _, f := range proxyStopListenPool {
+			f()
+		}
+	}
+	defer func() {
+		// On error, stop the servers before returning.
+		if err != nil {
+			stopServer()
+			stopServer = nil
+		}
+	}()
+
+	isBlockMigration := (vmi.Status.MigrationMethod == v1.BlockMigration)
+	migrationPortsRange := migrationproxy.GetMigrationPortsList(isBlockMigration)
+
+	// Create a tcp server for each direct connection proxy
+	for _, port := range migrationPortsRange {
+		key := migrationproxy.ConstructProxyKey(string(vmi.UID), port)
+		migrationProxy := migrationproxy.NewTargetProxy(loopbackAddress, port, nil, nil, migrationproxy.SourceUnixFile(virtShareDir, key))
+		err := migrationProxy.StartListening()
+		if err != nil {
+			return stopServer, err
+		}
```

Comment:
`err` is defined as a return argument in the function. Here you're defining `err` again within the scope of this for loop. In the defer function you're checking 

```
		if err != nil {
			stopServer()
			stopServer = nil
		}
```

due to the redefining of `err` at different scopes, it's unclear to me if the logic in that defer loop will do what you'd expect it to do when an error is encountered.

Given the complexity here, maybe this particular block shouldn't be refactored, and was easier to maintain in the top level function. Is that possible?

Reply:
I thought that this doesn't matter because we explicitly return the err 
But I understand that this may be confusing.

Reply:
> due to the redefining of `err` at different scopes, it's unclear to me if the logic in that defer loop will do what you'd expect it to do when an error is encountered.
> 
> Given the complexity here, maybe this particular block shouldn't be refactored, and was easier to maintain in the top level function. Is that possible?

I recognize that reusing the same name may be confusing, I can use a simple `e` in this case if it helps.
Regarding the complexity, I think this is a common pattern when using defers and it is used in the codebase especially when the deferred operation returns an error (like the famous file close).

The change itself started when I still did not know how the end result will look like, eventually this portion was not directly needed for the locking of the domain change, so we could remove it completely from this PR.

I can argue the following in its favor:
- The abstraction this change introduces is simplifying the caller significantly, but indeed adds some complexity in the helper.
- The original usage of the `defer` inside the for-loop is not less complex, it adds a "order" question mark (my IDE warns about using defers in loops).
- The usage of a named returned value is part of golang and is used on the codebase already.

Summarizing, if changing the name of the error variable will help, I would gladly favor that over dropping this change.
However, if you feel it is too much, I will drop the change.
Please let me know your preference on this.

Reply:
personally, the indirection in this helper function takes longer for me to grok than seeing everything laid out sequentially in the caller function. not overloading the err variable would make it more obvious. returning the logic to the caller would make it more obvious too. whatever you want to do.

Reply:
Thank you for freedom to choose one way or the other.
I prepared both options and went back and forth between them. I feel more comfortable with abstracting the layer of the proxy preparation, allowing the reader to choose if to delve into the details (about the how) or not.

I changed the error variable name and scoped it in the `if` statement.